### PR TITLE
Multi turn rollout

### PR DESCRIPTION
**Pull request by tastelikefeet**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

## Features
1. Add twinkle_agent package to hold multi-turn rollouts and tool callings
2. Add notifier to notify user when training is failed
3. Add GRPO entropy loss/ref logps loss
4. Add GRPO series metrics
5. Add a patch to qwen3 series models to fix a bug that special tokens will cause jinja encode error
6. Support sample with base model when enable_lora is True
7. Support lora model id used in sampling
8. Support Qwen models tool parsing and cleaning
9. Support selective_log_softmax returns entropy
10. Support api/vllm multi turn rollouts and trace saving
11. Support tool manager
12. [Experimental] Support passage chunker and condenser and tool

## Bug Fix
1. Fix dataset multi-process preprocessing
2. Fix a bug that mm model templates cause cache failed

## Experiment results

Paste your experiment result here(if needed).

### CODE REVIEW

File: src/twinkle/loss/grpo.py [hunk 1]
```diff
@@ -390,7 +382,9 @@ def _compute_per_token_loss(
     ) -> 'torch.Tensor':
         """Clamped ratio * advantage * log_prob."""
         import torch
-        clamped_ratios = torch.clamp(ratio, max=1 + self.epsilon).detach()
+
+        # Two-sided IS clamp with asymmetric epsilon, matching MiniMax CISPO spec.
+        clamped_ratios = torch.clamp(ratio, min=1 - self.epsilon, max=1 + self.epsilon_high).detach()
```
Comment (hjh0119) [inline src/twinkle/loss/grpo.py]: check it

File: src/twinkle/utils/torch_utils.py [hunk 2]
```diff
@@ -74,35 +74,65 @@ def selective_log_softmax(logits, index) -> 'torch.Tensor':
...
             row_per_token_logps = row_logps.gather(dim=-1, index=row_labels.unsqueeze(-1)).squeeze(-1)
             per_token_logps.append(row_per_token_logps)
+            if return_entropy:
+                # row_logps is already stable; softmax reuses the same numerics.
+                row_p = torch.exp(row_logps)
+                per_token_entropy.append(-(row_p * row_logps).sum(dim=-1))
         per_token_logps = torch.stack(per_token_logps)
+        if return_entropy:
+            return per_token_logps, torch.stack(per_token_entropy)
     return per_token_logps
```
Comment (hjh0119) [inline src/twinkle/utils/torch_utils.py:136]: Suggest unifying the return type: when `return_entropy=False`, 
return `(per_token_logps, None)` instead of just `per_token_logps`.

**Review by hjh0119 — APPROVED** [review]

File: src/twinkle_agentic/rollout/api_multi_turn.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        self.api = api
+        self.tool_manager = tool_manager
+        self.sampling_params = sp
+        self.max_turns = max_turns
+        self.concurrency = concurrency
+        self.extra_body = dict(extra_body or {})
+        self.trace_dir = trace_dir
+        self.trace_callback = trace_callback
+        self.success_callback = success_callback
+        os.makedirs(self.trace_dir, exist_ok=True)
```
Comment (tpx818) [inline src/twinkle_agentic/rollout/api_multi_turn.py]: 这个操作是否放在call中if self.trace_dir后处理

File: src/twinkle/utils/torch_utils.py (hunk 2 above)
Comment (tastelikefeet) [reply src/twinkle/utils/torch_utils.py:136]: Ideally yes, but selective_log_softmax has been used in too many cases, change the default return value format will cause unexpected errors
